docs/fetching_blocks.md

```diff
@@ -0,0 +1,159 @@
+   chain if it checks out.
+
+## A block store
+
+This is held in `block_store.rs`
+
+The fundamental ops are
+
+### `request_missing_blocks()`
+
+Which
+
+ - Looks through the blocks we have
+ - Finds the next blocks it thinks we need
+ - Iterates through our known peers.
+
+If we don't have availability for a peer, we will request it by
+sending an empty block request.
+
+If we do, we will try to request whatever blocks it has that we want.
+
+We limit the number of outstanding requests per peer, in order to
+avoid bufferbloat at the peer's input message queue.
+
+We don't ask for blocks that we think are in flight (ie. we've
+requested them but they have not yet arrived), those we don't think a
+peer has, or those we think are gaps (remember that requests are made
+by view, so you can't guarantee that every view has a block).
+
+We time out outstanding requests on a flat-timeout basis (our model
+being that if you haven't replied by now, the whole message has
+probably been lost).
+
+### `process_block()`
+
+Which, called originally from `tick()`, given a block, will attempt to
+process it. If it succeeded, it will try to retrieve the next block
+from the cache and send a `ProcessProposal` message to process it.
+
+We do this to avoid blocking `tick()` for too long.
+
+## Block arrival
+
+When blocks arrive (via `node.rs`'s `BlockResponse` handler), we put
+them directly in the block store's buffer cache. The tick will then
+process them next time it runs.
+
+When new proposals arrive (from us via `process_block()`) via the
+`ProcessProposal` message, we will attempt to process them. If this
+results in a next proposal to process from the block buffer cache, we
+will send another `ProcessProposal` message to process that one too.
+
+## Forks
+
+Forks are problematic. In particular, they break the flow of
+parent-child hashes, so we have to go hunting for blocks in our buffer
+which have parent hashes which our database knows about - we do this
+using a progressive additive range search to avoid penalising our
+database too much (see `fork_counter` in `block_store.rs`).
+
+We may also never pick up a fork. Suppose the chain forked at view
+202, giving a left hand block at 202 and a right hand block at 204
+which then became the rest of the chain.
+
+Now suppose we get 204, but never 202. How would we ever know that 202
+existed? Well, we might not and will thus never store it in our
+database. Be aware of this.
+
+Also, the maximum depth of a fork is the maximum lookahead of our
+cache. Suppose we have a fork 100 deep starting at view 202. We'll
+fetch it, go to the end, and stop.
+
+How would we ever know that there was another block whose parent was
+the block at view 200? Well, unless we ask for subsequent blocks, we
+won't. And we won't ask for subsequent blocks because we're too busy
+re-requesting the ones we can't process.
+
+This is theoretically fixable and we should one day - but I've not
```

Comment:
Only if the node that's telling you about them is honest. If it's lying, then it can construct forks of arbitrary depth and if they are long enough, you may not be able to see past them to pick up the true chain again :-(